### ISSUE TITLE
f2py cannot handle array PARAMETER

### ISSUE THREAD

**Issue by maxnoe** [numpy/numpy#11612]
Consider the following minimal example:
```fotran
module test
  implicit none
  real(8), parameter :: c(2) = (/ 1.0, 2.0 /)
end module test
```

using f2py succeeds, but in the log there is:

```
In: :test:test.f90:constants
get_parameters: got "invalid syntax (<string>, line 1)" on '(/1.0, 2.0/)'
```

**Comment by JohnWChapman** [comment]
Has this issue been addressed? I am also encountering the same issue when using array parameter.

**Comment by melissawm** [comment]
Hi! I can confirm this; the error message persists although the array parameter values are correctly read when the module is imported. However, I found an example where this fails:

```fortran
module test
  implicit none
  integer, parameter, dimension(2) :: myparamarray = (/ 3, 5 /)
  real(8) :: x, y
  dimension x(myparamarray(1)), y(myparamarray(2))
end module test
```

Compiling with f2py gives the following output (including everything for completeness):

```bash
$ f2py -c test_11612_useparameter.f90 -m test
running build
running config_cc
unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
build_src
building extension "test" sources
f2py options: []
f2py:> /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.c
creating /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7
Reading fortran codes...
        Reading file 'test_11612_useparameter.f90' (format:free)
Post-processing...
        Block: test
                        Block: test
In: :test:test_11612_useparameter.f90:test
get_parameters: got "invalid syntax (<string>, line 1)" on '(/3, 5/)'
Post-processing (stage 2)...
        Block: test
                Block: unknown_interface
                        Block: test
Building modules...
        Building module "test"...
                Constructing F90 module support for "test"...
                  Variables: myparamarray x y
        Wrote C/API module "test" to file "/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.c"
        Fortran 90 wrappers are saved to "/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/test-f2pywrappers2.f90"
  adding '/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/fortranobject.c' to sources.
  adding '/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7' to include_dirs.
copying /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.c -> /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7
copying /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.h -> /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7
  adding '/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/test-f2pywrappers2.f90' to sources.
build_src: building npy-pkg config files
running build_ext
customize UnixCCompiler
customize UnixCCompiler using build_ext
get_default_fcompiler: matching types: '['gnu95', 'intel', 'lahey', 'pg', 'absoft', 'nag', 'vast', 'compaq', 'intele', 'intelem', 'gnu', 'g95', 'pathf95', 'nagfor']'
customize Gnu95FCompiler
Found executable /usr/bin/gfortran
customize Gnu95FCompiler
customize Gnu95FCompiler using build_ext
building 'test' extension
compiling C sources
C compiler: gcc -pthread -B /opt/miniconda3/envs/numpy-dev/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC

creating /tmp/tmpk2lb3ip2/tmp
creating /tmp/tmpk2lb3ip2/tmp/tmpk2lb3ip2
creating /tmp/tmpk2lb3ip2/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7
compile options: '-I/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7 -I/opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include -I/opt/miniconda3/envs/numpy-dev/include/python3.7m -c'
gcc: /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.c
gcc: /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/fortranobject.c
In file included from /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1823,
                 from /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/fortranobject.h:13,
                 from /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/fortranobject.c:2:
/opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
In file included from /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1823,
                 from /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/fortranobject.h:13,
                 from /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.c:15:
/opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.c:90:13: error: expected expression before ‘/’ token
   90 |   {"x",1,{{(/3, 5/)(1)}},NPY_DOUBLE},
      |             ^
/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.c:91:13: error: expected expression before ‘/’ token
   91 |   {"y",1,{{(/3, 5/)(2)}},NPY_DOUBLE},
      |             ^
error: Command "gcc -pthread -B /opt/miniconda3/envs/numpy-dev/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7 -I/opt/miniconda3/envs/numpy-dev/lib/python3.7/site-packages/numpy/core/include -I/opt/miniconda3/envs/numpy-dev/include/python3.7m -c /tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.c -o /tmp/tmpk2lb3ip2/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.o -MMD -MF /tmp/tmpk2lb3ip2/tmp/tmpk2lb3ip2/src.linux-x86_64-3.7/testmodule.o.d" failed with exit status 1
```

What I don't know is: do we want this to work? 

@pearu , can you help decide if this should be done? if so, I can try to find a solution.

**Comment by pearu** [comment]
@melissawm The important error message is this:
```
get_parameters: got "invalid syntax (<string>, line 1)" on '(/3, 5/)'
```
(the rest is just a consequence). To fix it, I suggest the following strategy:
1. Fix `get_parameters` (in numpy/f2py/crackfortran.py) so that, say, `get_parameters(dict(a='(/3, 5/)'))` results in `dict(a=(3, 5))`
2. Observe what will be the next failure (I cannot predict the exact failure atm but it's likely that there will be one) and act accordingly.

**Comment by pearu** [comment]
To answer your first question: the `get_parameters` failure is irrelevant for the example in issue description (it is too minimal as the parameter `c` is not exposed to Python) but relevant for your example case where `myparamarray` determines the sizes of `x` and `y` which are exposed to Python.

**Comment by melissawm** [comment]
Ok, I managed to do an improvised fix. I'm sure this is not robust or pretty, but it works. I'm submitting it over at #15457 but if you have ideas on how to improve it, please let me know.

**Comment by HaoZeke** [comment]
This error is present on `main`, at https://github.com/numpy/numpy/commit/7f682bca564a15b631632f9074d3277499c297eb. Fixed by #15457.
